### PR TITLE
Tagging for replicated event sourcing

### PR DESCRIPTION
**Pull request by johanandren**
First steps of projections for replicated event sourcing, making sure tagging works

References #29312

### CODE REVIEW

File: akka-docs/src/main/paradox/typed/replicated-eventsourcing.md [hunk 1]
```diff
@@ -271,6 +271,43 @@ Java
...
+
+Scala
+:  @@snip [ReplicatedEventSourcingTaggingSpec.scala](/akka-persistence-typed-tests/src/test/scala/docs/akka/persistence/typed/ReplicatedEventSourcingTaggingSpec.scala) { #tagging }
+
+Java
+:  @@snip [ReplicatedStringSet.java](/akka-persistence-typed-tests/src/test/java/jdocs/akka/persistence/typed/ReplicatedStringSet.java) { #tagging }
+
+FIXME then do a more advanced one rather than the other way around 
+While the above example showcases various ways of tagging, the most common and simple way to tag events is to tag them 
+using the entity type name that is persisting these events, and doing so only in the origin datacenter, which is as simple as:
```
Comment (johanandren) [inline akka-docs/src/main/paradox/typed/replicated-eventsourcing.md:309]: not quite done with the copy pasta here

File: akka-persistence-typed-tests/src/test/scala/akka/persistence/typed/ReplicatedEventSourcingTaggingSpec.scala [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+      /* FIXME can't verify tags with testkit ;(
+      val query =
+        PersistenceQuery(system).readJournalFor[CurrentEventsByTagQuery](PersistenceTestKitReadJournal.Identifier)
+
+      val stringTaggedEvents = query.currentEventsByTag("strings", NoOffset).runWith(Sink.seq).futureValue
+      stringTaggedEvents.map(_.event).toSet should equal(allEvents)
+
+      val longStrings = query.currentEventsByTag("long-strings", NoOffset).runWith(Sink.seq).futureValue
+      longStrings should have size (1)
+     */
```
Comment (johanandren) [inline akka-persistence-typed-tests/src/test/scala/akka/persistence/typed/ReplicatedEventSourcingTaggingSpec.scala:108]: need to actually verify tags were written here, so I guess I need to use some other journal than the testkit or add tagging support to the testkit.

File: akka-docs/src/main/paradox/typed/replicated-eventsourcing.md [hunk 3]
```diff
@@ -271,6 +271,42 @@ Java
 
 More advanced routing among the replicas is currently left as an exercise for the reader (or may be covered in a future release [#29281](https://github.com/akka/akka/issues/29281), [#29319](https://github.com/akka/akka/issues/29319)).
 
+## Tagging events and running projections
+
+Just like for regular `EventSourcedBehavior`s it is possible to “tag” events along with persisting them.
```
Comment (chbatey) [inline akka-docs/src/main/paradox/typed/replicated-eventsourcing.md:276]: don't think we need quotes around tag

File: akka-docs/src/main/paradox/typed/replicated-eventsourcing.md [hunk 4]
```diff
@@ -271,6 +271,42 @@ Java
 
 More advanced routing among the replicas is currently left as an exercise for the reader (or may be covered in a future release [#29281](https://github.com/akka/akka/issues/29281), [#29319](https://github.com/akka/akka/issues/29319)).
 
+## Tagging events and running projections
+
+Just like for regular `EventSourcedBehavior`s it is possible to “tag” events along with persisting them. 
+This is useful for later retrival of events for a given tag. The same @ref[API for tagging provided for EventSourcedBehavior](persistence.md#tagging) can 
+be used for replicated event sourced behaviors as well.
+Tagging is useful in practice to build queries that lead to other data representations or aggregations of the these event 
+streams that can more directly serve user queries – known as building the “read side” in CQRS based applications.
```
Comment (chbatey) [inline akka-docs/src/main/paradox/typed/replicated-eventsourcing.md:280]: Link the CQRS page?

File: akka-docs/src/main/paradox/typed/replicated-eventsourcing.md [hunk 5]
```diff
@@ -271,6 +271,42 @@ Java
...
+The tagging is invoked in each replicas, which requires some special care in using tags, or else the same event will be
+tagged one time for each replica and show up in the event by tag stream one time for each replica. In addition to this
+the tags will be written in the respective journal of the replicas, which means that unless they all share a single journal
+the tag streams will be local to the replica even if the same tag is used on multiple replicas.
+
+One strategy for dealing with this is to include the replica id in the tag name, this means there will be a tagged stream of events
+per replica that contains all replicated events, but since the events can arrive in different order, they can also come in different
+order per replica tag.
+
+Another strategy would be to tag only the events that is local to the replica and not events that are replicated. Either
```
Comment (chbatey) [inline akka-docs/src/main/paradox/typed/replicated-eventsourcing.md:294]: are local

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala [hunk 6]
```diff
@@ -419,17 +412,20 @@ private[akka] object Running {
           updatedVersion,
           isConcurrent)
 
-      val newState: RunningState[S] = withContext(
-        replication,
-        aa => aa.setContext(recoveryRunning = false, event.originReplica, concurrent = isConcurrent),
-        () => state.applyEvent(setup, event.event))
+      replication.setContext(recoveryRunning = false, event.originReplica, concurrent = isConcurrent)
```
Comment (chbatey) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala:415]: 👍

**Review by chbatey — APPROVED** [review]
Looks good! Can't hurt to have both options

File: akka-docs/src/main/paradox/typed/replicated-eventsourcing.md (hunk 3 above)
Comment (johanandren) [reply akka-docs/src/main/paradox/typed/replicated-eventsourcing.md:276]: you don't "think" we do?

Comment (johanandren) [reply akka-docs/src/main/paradox/typed/replicated-eventsourcing.md:276]: I "agree"

**Pull request merged**